Add str_strict parameter type instead of warning by default on string conversion

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
To alleviate much of the confusion around the string conversion warning intrudoced in 2.8.0 (f52a088862), change the string type behavior back to the way it behaved prior to 2.8.0 (convert and don't warn).

Give module auhors a new type, `str_strict`, they can use to enforce strict string types.

Related to discussions in #56788 and https://github.com/ansible/ansible/issues/60335#issuecomment-520894152.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/config/base.yml`
`lib/ansible/module_utils/basic.py`
`test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py`
##### ADDITIONAL INFORMATION
The string conversion warning was added in Ansible 2.8.0 (f52a088862). Since then, it has caused more confusion and issues than any benefit it has provided (I should have [listened to Toshio](https://github.com/ansible/ansible/issues/50503#issuecomment-452455543) and many others that considered the automatic conversion a feature not a bug  😄 ).

Further complicating things is the fun number of hoops we go through in the `choices` evaluation. We run `check_argument_types()` before we run `_check_argument_values()`. In `_check_argument_values()` we look at the value of `choices` for a module and convert stringified booleans to `'yes'` or `'no'` strings if those values are in the `choices` list. This is a longstanding feature of Ansible and we should not mess with this behavior since it is helpful.

This PR effectively moves the current behavior of warning, erroring, or ignoring on string conversion to the `str_strict` type and restores the previos behavior for `str` parameters.

samdoran commented:
I'm not sure if other things need to be done to make this work for Windows modules as well.

jborean93 commented:
For Windows module we never added the warning in the first place so `type="str"` works the same there. We could add `str_strict` to match Python at some point in the future though.

jamescassell reviewed (approved):
shipit

This will be a great improvement for unnecessary warnings.

jamescassell commented:
Any chance of this getting backported to 2.8 (or 2.9) once it's merged?

samdoran commented:
@jamescassell Since this is introducing a new feature, it will not be backported.

samdoran committed:
Show 'str_strict' as 'str' in ansible-doc CLI output

samdoran committed:
Add changelog and porting guide

samdoran committed:
Add unit tests

samdoran committed:
Add str_strict parameter type

To alleviate much of the confusion around the string conversion warning intrudoced in 2.8.0, change the string type behavior back to the way it behaved prior to 2.8.0 (convert and don't warn).

Give module auhors a new type, str_strict, they can use to enforce strict string types.

samdoran committed:
Move porting guide entry

samdoran committed:
Tweak warning message

abadger commented on docs/docsite/rst/porting_guides/porting_guide_base_2.11.rst:
```diff
@@ -46,6 +46,15 @@ No notable changes
+
+A new parameter type, ``str_strict``, is now available. This allows module authors to have Ansible take the action specified in ``ANSIBLE_STRING_CONVERSION_ACTION`` (``warn``, ``error``, or ``ignore`` and silently convert to a string) when a value in a ``str_strict`` field is not a string, such as unquoted integer or boolean values.
+
+This is a change from what was introduced in 2.8 where this behavior applied to *all* ``str`` type parameters.
```
"introduced and later reverted" ?

samdoran replied:
Yes, that's accurate. Turns out automatically casting to a string is a feature of `str` not a bug.

felixfontein replied:
I would say it's both a feature **and** a bug. Unfortunately it's so deeply integrated in the way Ansible works that it's really hard (or even impossible) to get rid of.

abadger replied:
Oh, I see... what a complex history this one has...... Hmmm.... since this was introduced in 2.8 which is now several years in the past, do we want to introduce str_strict or do we want to introduce str_relaxed?

* Argument for str_strict: 
  *  We can revert the behaviour for every other use of str.  That way maintainers don't have to do anything for people to stop seeing the warning in the common case.
* Argument for str_relaxed:
  * Anyplace where it truly matters that string conversion will cause problems will need to be updated otherwise they will suddenly lose the automatic checking they were counting on to stop users from doing the wrong thing.

I see major drawbacks to either of these cases.  So I'm just going to leave this comment for your evaluation.

samdoran replied:
Judging by the issues raised since this change was made, most are opened due to the warning causing confusion and noise rather than identifying a useful problem.

I'm not aware of any cases where this brought to light a string conversion that would have silently caused a problem with the module execution. Doesn't mean there aren't any, though.

I'm in favor of preventing module maintainers from having to change their argument specs to get the pre-2.8 behavior of just casting `str` parameters to string.

felixfontein replied:
Well, I can just say that before the 2.8 change, there were several issues created for docker_container which were caused by unwanted string conversions. After the 2.8 change, I haven't seen any again (or at least right now I don't remember one). That's of course only anecdotical evidence, but I'm sure there was at least one case where this warning helped. Though probably less cases than where this was annoying and confusing :/

samdoran replied:
That's a good data point. I'm glad it has helped somewhat. :)

samdoran committed:
Fix bad merge

This effectively makes str_strict only available to module arg specs and not other plugin types
or role argument spec validation.

samdoran commented on lib/ansible/module_utils/basic.py:
```diff
@@ -743,6 +743,7 @@ def __init__(self, argument_spec, bypass_checks=False, no_log=False,
 
         # This is for backwards compatibility only.
         self._CHECK_ARGUMENT_TYPES_DISPATCHER = DEFAULT_TYPE_VALIDATORS
+        self._CHECK_ARGUMENT_TYPES_DISPATCHER['str_strict'] = self._check_type_str_strict
```
Since `DEFAULT_TYPE_VALIDATORS` moved out of `AnsibleModule`, I had to put this in to make `str_strict` work. I don't particularly like adding a private method to a constant where all other entries are standalone functions.

Since "strict string" checking is really just a call to `check_type_str()` with special handling, maybe this should move to a standalone function as well.

samdoran commented:
I've had some internal discussion about this fix and I'm leaning towards abandoning this PR for the time being as a result.

The main goal was to restore the original behavior of `type: 'str'` and keep this new behavior as an opt-in for module authors, removing the noisy and mostly useless warnings.

After looking at this behavior more broadly, the various `check_` methods are actually _coercion_, not just validation, methods since they mutate the input value if possible into the specified type. In the case of `dict` and `list`, we even accept a string with special formatting, `k1=v1,k2=v2`, and `1,2,3` respectively.

A couple of ideas discussed internally were making `<type>_strict` for all types or adding an `allow_conversion` parameter to `AnsibleModule` or for each argument in the spec. That would give module authors the ability to control the type coercion. There hasn't been huge demand for such a feature, so I'm not sure how worthwhile it would be to implement.

Either of those are probably a more complete solution than the band-aid this PR is adding.

felixfontein commented:
A `allow_conversion` parameter in argument_spec would be awesome, because it would be a backwards-compatible solution: existing Ansible versions would simply ignore it because they don't know about it - well, except that their validate-module sanity tests will complain.

This would allow collections which support Ansible 2.9 etc. to use this feature without doing crazy things.

About `allow_conversion`: I could image it being three-valued: `allow` (current behavior), `warn` (current behavior for `type=str` if not disabled by user), `deny` (directly call fail_json).

abadger commented:
A general allow_conversation in the argument spec might not be a good idea.  When Ansible does not get a value via yaml (specified at the command line or via k=v in a playbook) it deals in strings. So int, dict, list, etc are handed to the module as strings. Conversion is essential in those cases.  Module authors can't specify no-conversion in that general case without breaking those Ansible features. 

No-conversion specifically for str type is different from the general case. Controllerside Ansible will either get typed data (via yaml) or str (via commandline or k=v).  So if module side Ansible is expecting a str and received some other type, it can know for certain that the user specified the wrong type.

So type=strict_str might not be the right way.  But whatever you choose will need to limit itself to what to do when a string is expected module side.   Other types require conversion in order to work.

samdoran commented:
> A general allow_conversation in the argument spec might not be a good idea. 

Agreed. It would wreak havoc in a number of ways.

> So type=strict_str might not be the right way. But whatever you choose will need to limit itself to what to do when a string is expected module side. Other types require conversion in order to work.

Very good point. Maybe another option for `type=str` rather than another type would be the best way to handle this.

```python
'param': {'type': 'str', 'allow_conversion': False}
```
